File: mutate4go/mutation_catalog_test.go
package mutate4go

import (
	"os"
	"path/filepath"
	"testing"
)

func TestMutationCatalogDiscoversBooleanEqualityAndComparisonMutations(t *testing.T) {
	root := t.TempDir()
	file := filepath.Join(root, "sample.go")
	writeFile(t, file, `package demo

func truthy() bool {
	return true
}

func same(left int, right int) bool {
	return left == right
}

func larger(left int, right int) bool {
	return left > right
}

func smaller(left int, right int) bool {
	return left <= right
}
`)

	sites, err := MutationCatalog{}.Discover([]string{file})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertDescriptions(t, sites, []string{
		"replace true with false",
		"replace == with !=",
		"replace > with >=",
		"replace <= with <",
	})
}

func TestMutationCatalogIgnoresOperatorsInsideStringsAndComments(t *testing.T) {
	root := t.TempDir()
	file := filepath.Join(root, "literals.go")
	writeFile(t, file, `package demo

func text(left int, right int) string {
	_ = "true == false > <"
	// left == right > 0
	/* false != true <= >= */
	if left == right {
		return "same"
	}
	return "different"
}
`)

	sites, err := MutationCatalog{}.Discover([]string{file})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertDescriptions(t, sites, []string{"replace == with !="})
}

func TestMutationCatalogDiscoversArithmeticLogicalUnaryAndConstantMutations(t *testing.T) {
	root := t.TempDir()
	file := filepath.Join(root, "expanded.go")
	writeFile(t, file, `package demo

func add(left int, right int) int {
	return left + right
}

func divide(left int, right int) int {
	return left / right
}

func both(left bool, right bool) bool {
	return left && right
}

func invert(value bool) bool {
	return !value
}

func negative(value int) int {
	return -value
}

func zero() int {
	return 0
}

func one() int {
	return 1
}
`)

	sites, err := MutationCatalog{}.Discover([]string{file})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertDescriptions(t, sites, []string{
		"replace + with -",
		"replace / with *",
		"replace && with ||",
		"replace ! with removed !",
		"replace - with removed -",
		"replace 0 with 1",
		"replace 1 with 0",
	})
	if sites[3].ReplacementText != "" || sites[4].ReplacementText != "" {
		t.Fatalf("expected unary removals to use empty replacement text: %#v", sites)
	}
}

func TestMutationCatalogAnalyzeReturnsScopesAndModuleHash(t *testing.T) {
	root := t.TempDir()
	file := filepath.Join(root, "sample.go")
	writeFile(t, file, `package demo

func truthy() bool {
	return true
}

func same(left int, right int) bool {
	return left == right
}
`)

	analysis, err := MutationCatalog{}.Analyze(file)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if analysis.ModuleHash == "" {
		t.Fatal("expected module hash")
	}
	if len(analysis.Scopes) == 0 {
		t.Fatal("expected scopes")
	}
	if len(analysis.Sites) != 2 {
		t.Fatalf("unexpected site count: %d", len(analysis.Sites))
	}
}

func writeFile(t *testing.T, path string, content string) {
	t.Helper()
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatalf("write file: %v", err)
	}
}

func assertDescriptions(t *testing.T, sites []MutationSite, want []string) {
	t.Helper()
	if len(sites) != len(want) {
		t.Fatalf("unexpected site count %d, want %d: %#v", len(sites), len(want), sites)
	}
	for i := range want {
		if sites[i].Description != want[i] {
			t.Fatalf("site %d description = %q, want %q", i, sites[i].Description, want[i])
		}
	}
}

File: mutate4rs/tests/mutation_catalog_test.rs
use std::{
    fs,
    path::PathBuf,
    time::{SystemTime, UNIX_EPOCH},
};

use mutate4rs::MutationCatalog;

#[test]
fn discovers_boolean_equality_and_comparison_mutations() {
    let file = temp_rust_file(
        r#"fn truthy() -> bool {
    true
}

fn same(left: i32, right: i32) -> bool {
    left == right
}

fn larger(left: i32, right: i32) -> bool {
    left > right
}

fn smaller(left: i32, right: i32) -> bool {
    left <= right
}
"#,
    );

    let sites = MutationCatalog.discover(&[file.clone()]).unwrap();
    assert_eq!(
        vec![
            "replace true with false",
            "replace == with !=",
            "replace > with >=",
            "replace <= with <",
        ],
        sites
            .iter()
            .map(|site| site.description.clone())
            .collect::<Vec<_>>()
    );
}

#[test]
fn ignores_operators_inside_strings_and_comments() {
    let file = temp_rust_file(
        r#"fn sample(left: i32, right: i32) -> &'static str {
    let _text = "true == false > <";
    // left == right > 0
    if left == right {
        "same"
    } else {
        "different"
    }
}
"#,
    );

    let sites = MutationCatalog.discover(&[file]).unwrap();
    assert_eq!(
        vec!["replace == with !="],
        sites
            .iter()
            .map(|site| site.description.clone())
            .collect::<Vec<_>>()
    );
}

#[test]
fn discovers_arithmetic_logical_unary_and_constant_mutations() {
    let file = temp_rust_file(
        r#"fn add(left: i32, right: i32) -> i32 { left + right }
fn divide(left: i32, right: i32) -> i32 { left / right }
fn both(left: bool, right: bool) -> bool { left && right }
fn invert(value: bool) -> bool { !value }
fn negative(value: i32) -> i32 { -value }
fn zero() -> i32 { 0 }
fn one() -> i32 { 1 }
"#,
    );

    let sites = MutationCatalog.discover(&[file]).unwrap();
    assert_eq!(
        vec![
            "replace + with -",
            "replace / with *",
            "replace && with ||",
            "replace ! with removed !",
            "replace - with removed -",
            "replace 0 with 1",
            "replace 1 with 0",
        ],
        sites
            .iter()
            .map(|site| site.description.clone())
            .collect::<Vec<_>>()
    );
    assert_eq!("", sites[3].replacement_text);
    assert_eq!("", sites[4].replacement_text);
}

#[test]
fn analyze_returns_scopes_and_module_hash() {
    let file = temp_rust_file(
        r#"fn truthy() -> bool {
    true
}

fn same(left: i32, right: i32) -> bool {
    left == right
}
"#,
    );

    let analysis = MutationCatalog.analyze(&file).unwrap();
    assert!(!analysis.module_hash.is_empty());
    assert!(!analysis.scopes.is_empty());
    assert_eq!(2, analysis.sites.len());
}

fn temp_rust_file(content: &str) -> PathBuf {
    let unique = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    let dir = std::env::temp_dir().join(format!("mutate4rs-tests-{unique}"));
    fs::create_dir_all(&dir).unwrap();
    let file = dir.join("sample.rs");
    fs::write(&file, content).unwrap();
    file
}
